concepts/identity.md

```diff
@@ -0,0 +1,283 @@
+
+This ownership CAN be transferred to a different Ethereum address. Activation of this feature SHOULD be optional, and MUST include a user-modifiable grace period during which reverting to previous ownership is possible.
+
+A Meta Identity owner CAN perform transfer of ownership. A Meta Identity owner CAN appoint a number of Ethereum addresses to act as *custodians*. These CAN through a vote perform transfer of ownership. A vote MUST require a message signed by a respective private key. The *minimum number of addresses* that must make up a custodian role, and *minimum ratio of votes* required to execute a transfer CAN ONLY be determined by the Meta Identity owner.
+
+### Implementation
+
+
+The Meta Identity uses ENS as backend for storage and retrieval, interfaced by the Identity and IdentityResolver contracts.
+
+The Identity contract is upgradeable and MUST be registered in the IdentityController contract. As the IdentityController instantiates all permanent storage, and guarantees that the contents of these cannot be altered between upgrades. Therefore only the instance of the Identity contract that has been registered with the IdentityController can be considered valid for use.
+
+In the following we assume the META-ID to be `sha3("foo")`.
+
+#### Add a new identity
+
+```
+Identity.register(bytes32 metaidhash, address owner);
+```
+
+This will make `owner` the owner of the META-ID, along with the ENS node `namehash("foo.id.meta")`. If the META-ID has already been registered, the method will throw an exception.
+
+#### Add metadata for identity
+
+```
+IdentityResolver.setContent(bytes32 node, bytes32 hash);
+```
+
+sets the content hash `hash` as the metadata for the node `namehash("foo.id.meta")`
+
+#### Retrieve metadata for identity
+
+```
+IdentityResolver.content(bytes32 node);
+```
+
+gets the content hash of the metadata for the node namehash("foo.id.meta")
+
+#### Retrieve the current valid Identity contract
+
+```
+IdentityController.current();
+```
+
+gets the address of the current valid Identity Contract
+
+#### Not yet implemented
+
+* Identity Custody Scheme contracts
+* events for invalid attempts of changes to identity
+* events for initiated change identity owner vote
+* events for notifying identity monitors to request refreshing their indexing (used for updating public keys) ...
+
+### Metadata
+
+Meta Identity metadata MUST constitute two parts; Core and Auxiliary. Core metadata is REQUIRED and can be defined only once. The Auxiliary metadata is OPTIONAL, and an arbitrary number of Auxiliary metadata items may be defined.
+
+Metadata may ONLY be changed by the Meta Identity owner.
+
+#### Core
+
+The following properties are REQUIRED:
+
+```
+id		: META-ID // reverse point to META-ID this metadata belongs to
+name		: string // name whom identity represents
+parent		: CID // pointer to previous version of this metadata
```

Comment:
Given this is a required field, what happens if there is no parent?

Reply:
@nolash - not sure this has been implemented yet or what the value of `parent` is initialised to?

Reply:
@lmars I'd say 0x0?

Reply:
Well `0x0` is not a valid CID so following it would lead to an error. If it is there to indicate "there is no parent", then just leave it blank (and thus make it optional)?

Can you also clarify what it is used for?

Reply:
@lmars it's a linked list, `parent` pointing to the previous version of the metadata.

Actually, I would argue that 0x0 would be a valid *evaluated* cid, because all of the four CID fields allow effective value of 0x0;

* multibase 0x0: literal bytes
* cid v 0x0: backwards compatibility version
* multicodec 0x0: well, not literally listed, but the table is full of holes, many entries have mere `0x` so I'd say a value of `0x0` could be defended
* multihash 0x0: hash identifier `identity` is `0x0`, and hash value of `0x0` is valid.

In a pedantic perspective, the field is *not* optional if the intention is that we want to be able to follow versioning of the identity changes.

Reply:
> it's a linked list, parent pointing to the previous version of the metadata.

Yes I was more asking what is the purpose of this linked list. Why do I need to know the previous version of the metadata, and why is that required?

With respect to `0x0`, a CID is supposed to identify a piece of data. You are trying to use `0x0` to represent the absence of data, in which case you may as well just leave the field blank?

Reply:
Sorry, not a good answer earlier.

I don't have a particular usecase for when it would be useful to have an identity change history. Maybe revert in case of illegitimate updates, but I don't know how realistic that is.

Considering the importance of the identity information, and the relatively slight cost of having this pointer, I deemed it better to include it in case we would need it for something.

About the absence of data or not; pointers in programming languages are supposed to identify a piece of data, too, but it's perfectly legitimate to use null pointers to indicate the absence.